### PR TITLE
examine nil pointer before applying backend protocol configurations

### PR DESCRIPTION
**Pull request by freehan**
Mitigates #675

### CODE REVIEW

File: pkg/healthchecks/healthchecks.go [hunk 1]
```diff
@@ -62,6 +62,10 @@ const (
 	// backends, the port or named port specified in the Backend Service is
 	// used for health checking.
 	UseServingPortSpecification = "USE_SERVING_PORT"
+
+	newHealthCheckErrorMessageTemplate = "the %v health check configuration on the exiting health check %v is nil. " +
+		"This is usually caused by application protocol changed. " +
```
Comment (rramkumar1) [inline pkg/healthchecks/healthchecks.go]: what do you mean by exiting?

Comment (rramkumar1) [inline pkg/healthchecks/healthchecks.go]: "..caused by an application..."

File: pkg/healthchecks/healthchecks.go [hunk 2]
```diff
@@ -62,6 +62,10 @@ const (
...
+		"Please revert the change on application protocol to avoid this error message."
```
Comment (rramkumar1) [inline pkg/healthchecks/healthchecks.go]: "...revert the change on the application..."

File: pkg/healthchecks/healthchecks.go (hunk 1 above)
Comment (cmluciano) [inline pkg/healthchecks/healthchecks.go]:
```suggestion
		"This is usually caused by a change of the application's protocol." +
```

File: pkg/healthchecks/healthchecks.go (hunk 2 above)
Comment (cmluciano) [reply pkg/healthchecks/healthchecks.go]:
```suggestion
		"Please revert the change on the application's protocol to avoid this error message."
```

I don't have a lot of insight into how the ingress-gce works internally, but I think a TODO on handling this type of change would be helpful. I think progressing from HTTP-1.1 to HTTP-2 (soon HTTP-3) or introducing changes on the protocol to adopt a service mesh will be a common occurrence.

Comment (freehan) [reply pkg/healthchecks/healthchecks.go]: yeah. Agreed. In the long run we want to support this. But currently, due to GCE api changes, it was broken. So we need to put this short term fix in place so that it can be cherry picked.

**Comment by freehan** [comment]
fixed.

**Comment by mariantalla** [comment]
Hi, is there anything missing for this to be `/lgtm`ed?
cc @MrHohn @thockin

**Comment by mariantalla** [comment]
Assigning for `lgtm` (please reassign if I've done this wrong):
/assign @MrHohn

**Comment by MrHohn** [comment]
This LGTM, but will defer to @rramkumar1 to make the final call.
/assign @rramkumar1

File: pkg/healthchecks/healthchecks.go [hunk 3]
```diff
@@ -62,6 +62,10 @@ const (
 	// backends, the port or named port specified in the Backend Service is
 	// used for health checking.
 	UseServingPortSpecification = "USE_SERVING_PORT"
+
+	newHealthCheckErrorMessageTemplate = "the %v health check configuration on the existing health check %v is nil. " +
+		"This is usually caused by an application protocol changed. " +
```
Comment (bowei) [inline pkg/healthchecks/healthchecks.go]: can we give the name of the actual field involved, this error is a bit vague sounding. are we talking about the GCP object, the protocol used by the app itself, etc

File: pkg/healthchecks/healthchecks.go [hunk 4]
```diff
@@ -304,18 +311,30 @@ type HealthCheck struct {
...
 		v.HTTPHealthCheck = *hc.HttpHealthCheck
 	case annotations.ProtocolHTTPS:
+		if hc.HttpsHealthCheck == nil {
+			err = fmt.Errorf(newHealthCheckErrorMessageTemplate, annotations.ProtocolHTTPS, hc.Name)
+			break
+		}
 		v.HTTPHealthCheck = computealpha.HTTPHealthCheck(*hc.HttpsHealthCheck)
 	case annotations.ProtocolHTTP2:
+		if hc.Http2HealthCheck == nil {
+			err = fmt.Errorf(newHealthCheckErrorMessageTemplate, annotations.ProtocolHTTP2, hc.Name)
```
Comment (bowei) [inline pkg/healthchecks/healthchecks.go]: can we not return immediately (same in all case above)

File: pkg/healthchecks/healthchecks.go (hunk 3 above)
Comment (freehan) [reply pkg/healthchecks/healthchecks.go]: HealthCheck object contains HTTPHealthCheck configurations. It has bunch of fields. 

In this case, the controller thinks that HTTP2HealthCheck configuration is nil because it was not vendored.

**Comment by freehan** [comment]
Ready for next round

File: pkg/healthchecks/healthchecks.go [hunk 5]
```diff
@@ -62,6 +62,10 @@ const (
 	// backends, the port or named port specified in the Backend Service is
 	// used for health checking.
 	UseServingPortSpecification = "USE_SERVING_PORT"
+
+	newHealthCheckErrorMessageTemplate = "the %v health check configurations on the existing health check %v is nil. " +
```
Comment (rramkumar1) [inline pkg/healthchecks/healthchecks.go]: "the %v health check configuration..."

File: pkg/healthchecks/healthchecks.go (hunk 2 above)
Comment (rramkumar1) [reply pkg/healthchecks/healthchecks.go]: Can we add the TODO just so we remember to do it?

**Comment by freehan** [comment]
Done
